ipsets/ipsets.go

```diff
@@ -0,0 +1,810 @@
+	// atomically swap it into place.
+	mainSetName := ipSet.MainIPSetName
+	if !s.existingIPSetNames.Contains(mainSetName) {
+		// Create empty main IP set so we can share the atomic swap logic below.
+		// Note: we can't use the -exist flag (which should make the create idempotent)
+		// because it still fails if the IP set was previously created with different
+		// parameters.
+		s.logCxt.WithField("setID", ipSet.SetID).Debug("Pre-creating main IP set")
+		writeLine("create %s %s family %s maxelem %d",
+			mainSetName, ipSet.Type, s.IPVersionConfig.Family, ipSet.MaxSize)
+	}
+	tempSetName := ipSet.TempIPSetName
+	if s.existingIPSetNames.Contains(tempSetName) {
+		// Explicitly delete the temporary IP set so that we can recreate it with new
+		// parameters.
+		s.logCxt.WithField("setID", ipSet.SetID).Debug("Temp IP set exists, deleting it before rewrite")
+		writeLine("destroy %s", tempSetName)
+	}
+	// Create the temporary IP set with the current parameters.
+	writeLine("create %s %s family %s maxelem %d",
+		tempSetName, ipSet.Type, s.IPVersionConfig.Family, ipSet.MaxSize)
+	// Write all the members into the temporary IP set.
+	ipSet.pendingReplace.Iter(func(item interface{}) error {
+		member := item.(ipSetMember)
+		writeLine("add %s %s", tempSetName, member)
+		return nil
+	})
+	// Atomically swap the temporary set into place.
+	writeLine("swap %s %s", mainSetName, tempSetName)
+	// Then remove the temporary set (which was the old main set).
+	writeLine("destroy %s", tempSetName)
+
+	return
+}
+
+// writeDeltas calculates the ipset restore input required to apply the pending adds/deletes to the
+// main IP set.
+func (s *IPSets) writeDeltas(ipSet *ipSet, out io.Writer) (err error) {
+	mainSetName := ipSet.MainIPSetName
+	ipSet.pendingDeletions.Iter(func(item interface{}) error {
+		member := item.(ipSetMember)
+		_, err = fmt.Fprintf(out, "del %s %s --exist\n", mainSetName, member)
+		if err != nil {
+			return set.StopIteration
+		}
+		countNumIPSetLinesExecuted.Inc()
+		return nil
+	})
+	if err != nil {
+		return
+	}
+	ipSet.pendingAdds.Iter(func(item interface{}) error {
+		member := item.(ipSetMember)
+		_, err = fmt.Fprintf(out, "add %s %s\n", mainSetName, member)
+		if err != nil {
+			return set.StopIteration
+		}
+		countNumIPSetLinesExecuted.Inc()
+		return nil
+	})
+	return
+}
+
+// ApplyDeletions tries to delete any IP sets that are no longer needed.
+// Failures are ignored, deletions will be retried the next time AttemptCleanup() is called.
+func (s *IPSets) ApplyDeletions() {
+	s.pendingIPSetDeletions.Iter(func(item interface{}) error {
+		setName := item.(string)
+		logCxt := s.logCxt.WithField("setName", setName)
+		logCxt.Info("Deleting IP set (if it exists)")
+		if s.existingIPSetNames.Contains(setName) {
+			if err := s.deleteIPSet(setName); err != nil {
+				logCxt.WithError(err).Warning("Failed to delete IP set.")
+			}
+		}
+		return set.RemoveItem
+	})
+
+	s.gaugeNumIpsets.Set(float64(len(s.ipSetIDToIPSet)))
```

Comment:
Presumably `s.ipSetIDToIPSet` should be decremented in this code path, but I don't see where that is happening.